<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>DNA Sequence Statistics (1) &mdash; Bioinformatics 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="https://media.readthedocs.org/css/badge_only.css" type="text/css" />
    <link rel="stylesheet" href="https://media.readthedocs.org/css/readthedocs-doc-embed.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="https://media.readthedocs.org/javascript/jquery/jquery-2.0.3.min.js"></script>
    <script type="text/javascript" src="https://media.readthedocs.org/javascript/jquery/jquery-migrate-1.2.1.min.js"></script>
    <script type="text/javascript" src="https://media.readthedocs.org/javascript/underscore.js"></script>
    <script type="text/javascript" src="https://media.readthedocs.org/javascript/doctools.js"></script>
    <script type="text/javascript" src="https://media.readthedocs.org/javascript/readthedocs-doc-embed.js"></script>
    <link rel="top" title="Bioinformatics 0.1 documentation" href="https://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/index.html" />
    <link rel="next" title="DNA Sequence Statistics (2)" href="chapter2.html" />
    <link rel="prev" title="Neglected Tropical diseases" href="chapter0.html" />
 
<!-- RTD Extra Head -->



<!-- 
Read the Docs is acting as the canonical URL for your project. 
If you want to change it, more info is available in our docs:
  http://docs.readthedocs.org/en/latest/canonical.html
-->
<link rel="canonical" href="http://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/src/chapter1.html" />

<script type="text/javascript">
  // This is included here because other places don't have access to the pagename variable.
  var READTHEDOCS_DATA = {
    project: "a-little-book-of-r-for-bioinformatics",
    version: "latest",
    language: "en",
    page: "src/chapter1",
    builder: "sphinx",
    theme: "sphinxdoc",
    docroot: "/",
    source_suffix: ".rst",
    api_host: "https://readthedocs.org",
    commit: "8cd107b180ed964fe7c8021e78ff0ab528d5ba74"
  }
  // Old variables
  var doc_version = "latest";
  var doc_slug = "a-little-book-of-r-for-bioinformatics";
  var page_name = "src/chapter1";
  var html_theme = "sphinxdoc";
</script>
<!-- RTD Analytics Code -->
<!-- Included in the header because you don't have a footer block. -->
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-17997319-1']);
  _gaq.push(['_trackPageview']);

  // User Analytics Code
  _gaq.push(['user._setAccount', 'UA-36029021-1']);
  _gaq.push(['user._trackPageview']);
  // End User Analytics Code


  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<!-- end RTD Analytics Code -->
<!-- end RTD <extrahead> -->

  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="chapter2.html" title="DNA Sequence Statistics (2)"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="chapter0.html" title="Neglected Tropical diseases"
             accesskey="P">previous</a> |</li>
        <li><a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/index.html">Bioinformatics 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="chapter1.html#">DNA Sequence Statistics (1)</a><ul>
<li><a class="reference internal" href="chapter1.html#using-r-for-bioinformatics">Using R for Bioinformatics</a></li>
<li><a class="reference internal" href="chapter1.html#r-packages-for-bioinformatics-bioconductor-and-seqinr">R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li><a class="reference internal" href="chapter1.html#fasta-format">FASTA format</a></li>
<li><a class="reference internal" href="chapter1.html#the-ncbi-sequence-database">The NCBI sequence database</a></li>
<li><a class="reference internal" href="chapter1.html#retrieving-genome-sequence-data-via-the-ncbi-website">Retrieving genome sequence data via the NCBI website</a></li>
<li><a class="reference internal" href="chapter1.html#retrieving-genome-sequence-data-using-seqinr">Retrieving genome sequence data using SeqinR</a></li>
<li><a class="reference internal" href="chapter1.html#writing-sequence-data-out-as-a-fasta-file">Writing sequence data out as a FASTA file</a></li>
<li><a class="reference internal" href="chapter1.html#reading-sequence-data-into-r">Reading sequence data into R</a></li>
<li><a class="reference internal" href="chapter1.html#length-of-a-dna-sequence">Length of a DNA sequence</a></li>
<li><a class="reference internal" href="chapter1.html#base-composition-of-a-dna-sequence">Base composition of a DNA sequence</a></li>
<li><a class="reference internal" href="chapter1.html#gc-content-of-dna">GC Content of DNA</a></li>
<li><a class="reference internal" href="chapter1.html#dna-words">DNA words</a></li>
<li><a class="reference internal" href="chapter1.html#summary">Summary</a></li>
<li><a class="reference internal" href="chapter1.html#links-and-further-reading">Links and Further Reading</a></li>
<li><a class="reference internal" href="chapter1.html#acknowledgements">Acknowledgements</a></li>
<li><a class="reference internal" href="chapter1.html#contact">Contact</a></li>
<li><a class="reference internal" href="chapter1.html#license">License</a></li>
<li><a class="reference internal" href="chapter1.html#exercises">Exercises</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="chapter0.html"
                        title="previous chapter">Neglected Tropical diseases</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="chapter2.html"
                        title="next chapter">DNA Sequence Statistics (2)</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/_sources/src/chapter1.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="https://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="dna-sequence-statistics-1">
<h1>DNA Sequence Statistics (1)<a class="headerlink" href="chapter1.html#dna-sequence-statistics-1" title="Permalink to this headline">¶</a></h1>
<div class="section" id="using-r-for-bioinformatics">
<h2>Using R for Bioinformatics<a class="headerlink" href="chapter1.html#using-r-for-bioinformatics" title="Permalink to this headline">¶</a></h2>
<p>This booklet tells you how to use the R software to carry out some simple analyses
that are common in bioinformatics. In particular, the focus is on computational analysis
of biological sequence data such as genome sequences and protein sequences.</p>
<p>This booklet assumes that the reader has some basic knowledge of biology, but not
necessarily of bioinformatics. The focus of the booklet is to explain simple bioinformatics
analysis, and to explain how to carry out these analyses using R.</p>
<p>To use R, you first need to start the R program on your computer.
You should have already installed R on your computer (if not, for instructions on how to
install R, see <a class="reference external" href="installr.html">How to install R</a>).</p>
</div>
<div class="section" id="r-packages-for-bioinformatics-bioconductor-and-seqinr">
<h2>R packages for bioinformatics: Bioconductor and SeqinR<a class="headerlink" href="chapter1.html#r-packages-for-bioinformatics-bioconductor-and-seqinr" title="Permalink to this headline">¶</a></h2>
<p>Many authors have written R packages for performing a wide variety
of analyses. These do not come with the standard R installation,
but must be installed and loaded as &#8220;add-ons&#8221;.</p>
<p>Bioinformaticians have written several specialised <em>packages</em> for
R. In this practical, you will learn to use the SeqinR package to
retrieve sequences from a DNA sequence database, and to carry out
simple analyses of DNA sequences.</p>
<p>Some well known bioinformatics packages for R are the Bioconductor
set of R packages
(<a class="reference external" href="http://www.bioconductor.org/">www.bioconductor.org</a>), which
contains several packages with many R functions for analysing
biological data sets such as microarray data; and the SeqinR
package
(<a class="reference external" href="http://pbil.univ-lyon1.fr/software/seqinr/home.php?lang=eng">pbil.univ-lyon1.fr/software/seqinr/home.php?lang=eng</a>),
which contains R functions for obtaining sequences from DNA and protein
sequence databases, and for analysing DNA and protein sequences.</p>
<p>To use function from the SeqinR package,
we first need to install the SeqinR package (for instructions on how to
install an R package, see <a class="reference external" href="installr.html#how-to-install-an-r-package">How to install an R package</a>).
Once you have installed the &#8220;SeqinR&#8221; R package, you can load the &#8220;SeqinR&#8221; R package by typing:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> <span class="kn">library</span><span class="p">(</span><span class="s">&quot;seqinr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Remember that you can ask for more information about a particular R
command by using the help() function. For example, to ask for more
information about the library() function, you can type:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> help<span class="p">(</span><span class="s">&quot;library&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="fasta-format">
<h2>FASTA format<a class="headerlink" href="chapter1.html#fasta-format" title="Permalink to this headline">¶</a></h2>
<p>The FASTA format is a simple and widely used format for storing
biological (DNA or protein) sequences. It was first used by the
FASTA program for sequence alignment. It begins with a single-line
description starting with a &#8220;&gt;&#8221; character, followed by lines of
sequences. Here is an example of a FASTA file:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> A06852 <span class="m">183</span> residues
MPRLFSYLLGVWLLLSQLPREIPGQSTNDFIKACGRELVRLWVEICGSVSWGRTALSLEE
PQLETGPPAETMPSSITKDAEILKMMLEFVPNLPQELKATLSERQPSLRELQQSASKDSN
LNFEEFKKIILNRQNEAEDKSLLELKNLGLDKHSRKKRLFRMTLSEKCCQVGCIRKDIAR
LC
</pre></div>
</div>
</div>
<div class="section" id="the-ncbi-sequence-database">
<h2>The NCBI sequence database<a class="headerlink" href="chapter1.html#the-ncbi-sequence-database" title="Permalink to this headline">¶</a></h2>
<p>The National Centre for Biotechnology Information (NCBI)
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/">www.ncbi.nlm.nih.gov</a>) in the US
maintains a huge database of all the DNA and protein sequence data
that has been collected, the NCBI Sequence Database. This also a
similar database in Europe, the European Molecular Biology
Laboratory (EMBL) Sequence Database
(<a class="reference external" href="http://www.ebi.ac.uk/embl/">www.ebi.ac.uk/embl</a>), and also a
similar database in Japan, the DNA Data Bank of Japan (DDBJ;
<a class="reference external" href="http://www.ddbj.nig.ac.jp/">www.ddbj.nig.ac.jp</a>). These three
databases exchange data every night, so at any one point in time,
they contain almost identical data.</p>
<p>Each sequence in the NCBI Sequence Database is stored in a separate
<em>record</em>, and is assigned a unique identifier that can be used to
refer to that sequence record. The identifier is known as an
<em>accession</em>, and consists of a mixture of numbers and letters. For
example, Dengue virus causes <a class="reference external" href="http://apps.who.int/tdr/svc/diseases/dengue">Dengue fever</a>,
which is classified as a neglected tropical disease by the WHO.
by any one of four types of Dengue virus: DEN-1, DEN-2, DEN-3, and DEN-4.
The NCBI accessions for the DNA sequences of the DEN-1, DEN-2, DEN-3, and DEN-4
Dengue viruses are NC_001477, NC_001474, NC_001475 and NC_002640, respectively.</p>
<p>Note that because the NCBI Sequence Database, the EMBL Sequence
Database, and DDBJ exchange data every night, the DEN-1 (and DEN-2, DEN-3, DEN-4) Dengue virus
sequence will be present in all three databases, but it will
have different accessions in each database, as they each use their
own numbering systems for referring to their own sequence records.</p>
</div>
<div class="section" id="retrieving-genome-sequence-data-via-the-ncbi-website">
<h2>Retrieving genome sequence data via the NCBI website<a class="headerlink" href="chapter1.html#retrieving-genome-sequence-data-via-the-ncbi-website" title="Permalink to this headline">¶</a></h2>
<p>You can easily retrieve DNA or protein sequence data from the NCBI
Sequence Database via its website
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/">www.ncbi.nlm.nih.gov</a>.</p>
<p>The Dengue DEN-1 DNA sequence is a viral DNA sequence, and as
mentioned above, its NCBI accession is NC_001477. To retrieve
the DNA sequence for the Dengue DEN-1 virus from NCBI, go to the NCBI
website, type &#8220;NC_001477&#8221; in the Search box at the top of
the webpage, and press the &#8220;Search&#8221; button beside the Search box:</p>
<p><img alt="image0" src="../_images/P1_image0.png" /></p>
<p>On the results page you will see the number of hits to &#8220;NC_001477&#8221;
in each of the NCBI databases on the NCBI website. There are many
databases on the NCBI website, for example, the &#8220;PubMed&#8221; data
contains abstracts from scientific papers, the &#8220;Nucleotide&#8221;
database contains DNA and RNA sequence data, the &#8220;Protein&#8221; data
contains protein sequence data, and so on. The picture below shows
what the results page should look like for your NC_001477 search.
As you are looking for the DNA sequence of the Dengue DEN-1 virus
genome, you expect to see a hit in the NCBI Nucleotide database,
and indeed there is hit in the Nucleotide database (indicated by
the &#8220;1&#8221; beside the icon for the Nucleotide database):</p>
<p><a class="reference internal" href="../_images/P1_image1.png"><img alt="image1" src="../_images/P1_image1.png" style="width: 900px;" /></a></p>
<p>To look at the one sequence found in the Nucleotide database, you
need to click on the icon for the NCBI Nucleotide database on the
results page for the search:</p>
<p><img alt="image2" src="../_images/P1_image2.png" /></p>
<p>When you click on the icon for the NCBI Nucleotide database, it
will bring you to the record for NC_001477 in the NCBI Nucleotide
database. This will contain the name and NCBI accession of the
sequence, as well as other details such as any papers describing
the sequence:</p>
<p><img alt="image3" src="../_images/P1_image3.png" /></p>
<p>To retrieve the DNA sequence for the DEN-1 Dengue virus genome
sequence as a FASTA format sequence file, click on &#8220;Send&#8221; at the top
right of the NC_001477 sequence record webpage, and then choose
&#8220;File&#8221; in the pop-up menu that appears, and then choose FASTA
from the &#8220;Format&#8221; menu that appears, and click on &#8220;Create file&#8221;.</p>
<p>A box will pop up asking you what to name the file, and where to save it. You should give it a sensible
name (eg. &#8220;den1.fasta&#8221;) and save it in a place where you will
remember (eg. in the &#8220;My Documents&#8221; folder is a good idea):</p>
<p><img alt="image4" src="../_images/P1_image4.png" /></p>
<p>You can now open the FASTA file containing the DEN-1 Dengue virus genome
sequence using WordPad on your computer. To open WordPad, click on
&#8220;Start&#8221; on the bottom left of your screen, click on &#8220;All Programs&#8221;
in the menu that appears, and then select &#8220;Accessories&#8221; from the
menu that appears next, and then select &#8220;WordPad&#8221; from the menu
that appears next. WordPad should start up. In Wordpad, choose
&#8220;Open&#8221; from the &#8220;File&#8221; menu. The WordPad &#8220;Open&#8221; dialog will appear.
Set &#8220;Files of type&#8221; to &#8220;All Documents&#8221; at the bottom of the WordPad
&#8220;Open&#8221; dialog. You should see a list of files, now select the file
that contains the DEN-1 Dengue virus sequence (eg. &#8220;den1.fasta&#8221;). The
contents of the FASTA format file containing the Dengue DEN-1 sequence
should now be displayed in WordPad:</p>
<p><img alt="image5" src="../_images/P1_image5.png" /></p>
</div>
<div class="section" id="retrieving-genome-sequence-data-using-seqinr">
<h2>Retrieving genome sequence data using SeqinR<a class="headerlink" href="chapter1.html#retrieving-genome-sequence-data-using-seqinr" title="Permalink to this headline">¶</a></h2>
<p>Instead of going to the NCBI website to retrieve sequence data from the NCBI database, you
can retrieve sequence data from NCBI directly from R, by using the SeqinR R package.</p>
<p>For example, you learnt above how to retrieve the DEN-1 Dengue virus genome sequence,
which has NCBI accession NC_001477, from the NCBI website. To retrieve a sequence with
a particular NCBI accession, you can use R function &#8220;getncbiseq()&#8221; below, which you will
first need to copy and paste into R:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> getncbiseq <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>accession<span class="p">)</span>
  <span class="p">{</span>
     <span class="kn">require</span><span class="p">(</span><span class="s">&quot;seqinr&quot;</span><span class="p">)</span> <span class="c1"># this function requires the SeqinR R package</span>
     <span class="c1"># first find which ACNUC database the accession is stored in:</span>
     dbs <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;genbank&quot;</span><span class="p">,</span><span class="s">&quot;refseq&quot;</span><span class="p">,</span><span class="s">&quot;refseqViruses&quot;</span><span class="p">,</span><span class="s">&quot;bacterial&quot;</span><span class="p">)</span>
     numdbs <span class="o">&lt;-</span> <span class="kp">length</span><span class="p">(</span>dbs<span class="p">)</span>
     <span class="kr">for</span> <span class="p">(</span>i <span class="kr">in</span> <span class="m">1</span><span class="o">:</span>numdbs<span class="p">)</span>
     <span class="p">{</span>
        db <span class="o">&lt;-</span> dbs<span class="p">[</span>i<span class="p">]</span>
        choosebank<span class="p">(</span>db<span class="p">)</span>
        <span class="c1"># check if the sequence is in ACNUC database &#39;db&#39;:</span>
        resquery <span class="o">&lt;-</span> <span class="kp">try</span><span class="p">(</span>query<span class="p">(</span><span class="s">&quot;.tmpquery&quot;</span><span class="p">,</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;AC=&quot;</span><span class="p">,</span> accession<span class="p">)),</span> silent <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
        <span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="kp">inherits</span><span class="p">(</span>resquery<span class="p">,</span> <span class="s">&quot;try-error&quot;</span><span class="p">)))</span>
        <span class="p">{</span>
           queryname <span class="o">&lt;-</span> <span class="s">&quot;query2&quot;</span>
           thequery <span class="o">&lt;-</span> <span class="kp">paste</span><span class="p">(</span><span class="s">&quot;AC=&quot;</span><span class="p">,</span>accession<span class="p">,</span>sep<span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
           query<span class="p">(</span><span class="sb">`queryname`</span><span class="p">,</span><span class="sb">`thequery`</span><span class="p">)</span>
           <span class="c1"># see if a sequence was retrieved:</span>
           seq <span class="o">&lt;-</span> getSequence<span class="p">(</span>query2<span class="o">$</span>req<span class="p">[[</span><span class="m">1</span><span class="p">]])</span>
           closebank<span class="p">()</span>
           <span class="kr">return</span><span class="p">(</span><span class="kp">seq</span><span class="p">)</span>
        <span class="p">}</span>
        closebank<span class="p">()</span>
     <span class="p">}</span>
     <span class="kp">print</span><span class="p">(</span><span class="kp">paste</span><span class="p">(</span><span class="s">&quot;ERROR: accession&quot;</span><span class="p">,</span>accession<span class="p">,</span><span class="s">&quot;was not found&quot;</span><span class="p">))</span>
  <span class="p">}</span>
</pre></div>
</div>
<p>Once you have copied and pasted the function getncbiseq() into R, you can use it to retrieve
a sequence from the NCBI Nucleotide database, such as the sequence for the DEN-1 Dengue virus
(accession NC_001477):</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> dengueseq <span class="o">&lt;-</span> getncbiseq<span class="p">(</span><span class="s">&quot;NC_001477&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The variable <em>dengueseq</em> is a vector containing the nucleotide
sequence. Each element of the vector contains one nucleotide of the
sequence. Therefore, to print out a certain subsequence of the
sequence, we just need to type the name of the vector <em>dengueseq</em>
followed by the square brackets containing the indices for those
nucleotides. For example, the following command prints out the
first 50 nucleotides of the DEN-1 Dengue virus genome sequence:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> dengueseq<span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">50</span><span class="p">]</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;a&quot;</span>
<span class="p">[</span><span class="m">20</span><span class="p">]</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;a&quot;</span>
<span class="p">[</span><span class="m">39</span><span class="p">]</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;a&quot;</span> <span class="s">&quot;g&quot;</span> <span class="s">&quot;c&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;t&quot;</span> <span class="s">&quot;g&quot;</span>
</pre></div>
</div>
<p>Note that <em>dengueseq[1:50]</em> refers to the elements of the vector
<em>dengueseq</em> with indices from 1-50. These elements contain the
first 50 nucleotides of the DEN-1 Dengue virus sequence.</p>
</div>
<div class="section" id="writing-sequence-data-out-as-a-fasta-file">
<h2>Writing sequence data out as a FASTA file<a class="headerlink" href="chapter1.html#writing-sequence-data-out-as-a-fasta-file" title="Permalink to this headline">¶</a></h2>
<p>If you have retrieved a sequence from the NCBI database using the
&#8220;getncbiseq()&#8221; function, you may want to save the sequence to a FASTA-format
file on your computer, in case you need the sequence for further analyses (either
in R or in other software).</p>
<p>You can write out a sequence to a FASTA-format file in R by using the &#8220;write.fasta()&#8221;
function from the SeqinR R package. The write.fasta() function requires that you tell
it the name of the output file using the &#8220;file.out&#8221; argument (input). You also need
to specify the R variable that contains the sequence using the &#8220;sequences&#8221; argument,
and the name that you want to give to the sequence using the &#8220;names&#8221; argument.</p>
<p>For example, if you have stored the DEN-1 Dengue virus sequence in a vector <em>dengueseq</em>,
to write out the sequence to a FASTA-format file called &#8220;den1.fasta&#8221; that contains
the sequence labelled as &#8220;DEN-1&#8221;, you can type:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> write.fasta<span class="p">(</span>names<span class="o">=</span><span class="s">&quot;DEN-1&quot;</span><span class="p">,</span> sequences<span class="o">=</span>dengueseq<span class="p">,</span> file.out<span class="o">=</span><span class="s">&quot;den1.fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="reading-sequence-data-into-r">
<h2>Reading sequence data into R<a class="headerlink" href="chapter1.html#reading-sequence-data-into-r" title="Permalink to this headline">¶</a></h2>
<p>Using the SeqinR package in R, you can easily read a DNA sequence
from a FASTA file into R. For example, we described above how to
retrieve the DEN-1 Dengue virus genome sequence from the NCBI
database, or from R using the getncbiseq() function, and save it in a
FASTA format file (eg. &#8220;den1.fasta&#8221;).</p>
<p>You can read this FASTA format file into R using the read.fasta()
function from the SeqinR R package:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> <span class="kn">library</span><span class="p">(</span><span class="s">&quot;seqinr&quot;</span><span class="p">)</span>
<span class="o">&gt;</span> dengue <span class="o">&lt;-</span> read.fasta<span class="p">(</span>file <span class="o">=</span> <span class="s">&quot;den1.fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that R expects the files that you read in (eg. &#8220;den1.fasta&#8221;)
to be in the &#8220;My Documents&#8221; folder on your computer, so if you
stored &#8220;den1.fasta&#8221; somewhere else, you will have to move or copy
it into &#8220;My Documents&#8221;.</p>
<p>The command above reads the contents of the fasta format file
den1.fasta into an R object called <em>dengue</em>. The variable
<em>dengue</em> is an R list object. As explained above, a list is an R
object that is like a vector, but can contain elements that are
numeric and/or contain characters. In this case, the list <em>dengue</em>
contains information from the FASTA file that you have read in (ie.
the name given to the dengue sequence in the FASTA file, and the DNA sequence
itself). In fact, the first element of the list object <em>dengue</em>
contains the the DNA sequence. As described above, we can access
the elements of an R list object using double square brackets.
Thus, we can store the DNA sequence for DEN-1 Dengue virus in a
variable <em>dengueseq</em> by typing:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> dengueseq <span class="o">&lt;-</span> dengue<span class="p">[[</span><span class="m">1</span><span class="p">]]</span>
</pre></div>
</div>
<p>Now the variable <em>dengueseq</em> is a vector containing the nucleotide
sequence.</p>
</div>
<div class="section" id="length-of-a-dna-sequence">
<h2>Length of a DNA sequence<a class="headerlink" href="chapter1.html#length-of-a-dna-sequence" title="Permalink to this headline">¶</a></h2>
<p>Once you have retrieved a DNA sequence, we can obtain some simple
statistics to describe that sequence, such as the sequence&#8217;s total
length in nucleotides. In the above example, we retrieved the
DEN-1 Dengue virus genome sequence, and stored it in the vector
variable <em>dengueseq</em>. To subsequently obtain the length of the
genome sequence, we would use the length() function, typing:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> <span class="kp">length</span><span class="p">(</span>dengueseq<span class="p">)</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">10735</span>
</pre></div>
</div>
<p>The length() function will give you back the length of the sequence
stored in variable <em>dengueseq</em>, in nucleotides. The length()
function actually gives the number of elements in the input vector
that you pass to it, which in this case in the number of elements
in the vector <em>dengueseq</em>. Since each element of the vector
<em>dengueseq</em> contains one nucleotide of the DEN-1 Dengue virus
sequence, the result for the DEN-1 Dengue virus genome tells us
the length of its genome sequence (ie. 10735 nucleotides long).</p>
</div>
<div class="section" id="base-composition-of-a-dna-sequence">
<h2>Base composition of a DNA sequence<a class="headerlink" href="chapter1.html#base-composition-of-a-dna-sequence" title="Permalink to this headline">¶</a></h2>
<p>An obvious first analysis of any DNA sequence is to count the
number of occurrences of the four different nucleotides (&#8220;A&#8221;, &#8220;C&#8221;,
&#8220;G&#8221;, and &#8220;T&#8221;) in the sequence. This can be done using the the
table() function. For example, to find the number of As, Cs, Gs,
and Ts in the DEN-1 Dengue virus sequence (which you have put
into vector variable <em>dengueseq</em>, using the commands above), you
would type:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> <span class="kp">table</span><span class="p">(</span>dengueseq<span class="p">)</span>
dengueseq
  a    <span class="kt">c</span>    g    <span class="kp">t</span>
<span class="m">3426</span> <span class="m">2240</span> <span class="m">2770</span> <span class="m">2299</span>
</pre></div>
</div>
<p>This means that the DEN-1 Dengue virus genome sequence has 3426
As, 2240 Cs, 2770 Gs and 2299 Ts.</p>
</div>
<div class="section" id="gc-content-of-dna">
<h2>GC Content of DNA<a class="headerlink" href="chapter1.html#gc-content-of-dna" title="Permalink to this headline">¶</a></h2>
<p>One of the most fundamental properties of a genome sequence is its
GC content, the fraction of the sequence that consists of Gs and
Cs, ie. the %(G+C).</p>
<p>The GC content can be calculated as the percentage of the bases in
the genome that are Gs or Cs. That is, GC content = (number of Gs +
number of Cs)*100/(genome length). For example, if the genome is
100 bp, and 20 bases are Gs and 21 bases are Cs, then the GC
content is (20 + 21)*100/100 = 41%.</p>
<p>You can easily calculate the GC content based on the number of As,
Gs, Cs, and Ts in the genome sequence. For example, for the
DEN-1 Dengue virus genome sequence, we know from using the
table() function above that the genome contains 3426 As, 2240 Cs,
2770 Gs and 2299 Ts. Therefore, we can calculate the GC content
using the command:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> <span class="p">(</span><span class="m">2240+2770</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="o">/</span><span class="p">(</span><span class="m">3426+2240+2770+2299</span><span class="p">)</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">46.66977</span>
</pre></div>
</div>
<p>Alternatively, if you are feeling lazy, you can use the GC()
function in the SeqinR package, which gives the fraction of bases
in the sequence that are Gs or Cs.</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> GC<span class="p">(</span>dengueseq<span class="p">)</span>
<span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">0.4666977</span>
</pre></div>
</div>
<p>The result above means that the fraction of bases in the
DEN-1 Dengue virus genome that are Gs or Cs is 0.4666977. To
convert the fraction to a percentage, we have to multiply by 100,
so the GC content as a percentage is 46.66977%.</p>
</div>
<div class="section" id="dna-words">
<h2>DNA words<a class="headerlink" href="chapter1.html#dna-words" title="Permalink to this headline">¶</a></h2>
<p>As well as the frequency of each of the individual nucleotides
(&#8220;A&#8221;, &#8220;G&#8221;, &#8220;T&#8221;, &#8220;C&#8221;) in a DNA sequence, it is also interesting to
know the frequency of longer DNA &#8220;words&#8221;. The individual
nucleotides are DNA words that are 1 nucleotide long, but we may
also want to find out the frequency of DNA words that are 2
nucleotides long (ie. &#8220;AA&#8221;, &#8220;AG&#8221;, &#8220;AC&#8221;, &#8220;AT&#8221;, &#8220;CA&#8221;, &#8220;CG&#8221;, &#8220;CC&#8221;,
&#8220;CT&#8221;, &#8220;GA&#8221;, &#8220;GG&#8221;, &#8220;GC&#8221;, &#8220;GT&#8221;, &#8220;TA&#8221;, &#8220;TG&#8221;, &#8220;TC&#8221;, and &#8220;TT&#8221;), 3
nucleotides long (eg. &#8220;AAA&#8221;, &#8220;AAT&#8221;, &#8220;ACG&#8221;, etc.), 4 nucleotides
long, etc.</p>
<p>To find the number of occurrences of DNA words of a particular
length, we can use the count() function from the R SeqinR package. For example, to find
the number of occurrences of DNA words that are 1 nucleotide long
in the sequence <em>dengueseq</em>, we type:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> count<span class="p">(</span>dengueseq<span class="p">,</span> <span class="m">1</span><span class="p">)</span>
  a    <span class="kt">c</span>    g    <span class="kp">t</span>
 <span class="m">3426</span> <span class="m">2240</span> <span class="m">2770</span> <span class="m">2299</span>
</pre></div>
</div>
<p>As expected, this gives us the number of occurrences of the
individual nucleotides. To find the number of occurrences of DNA
words that are 2 nucleotides long, we type:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> count<span class="p">(</span>dengueseq<span class="p">,</span> <span class="m">2</span><span class="p">)</span>
  aa   ac   ag   at   ca   cc   cg   ct   ga   gc   gg   gt   ta   tc   tg   tt
 <span class="m">1108</span>  <span class="m">720</span>  <span class="m">890</span>  <span class="m">708</span>  <span class="m">901</span>  <span class="m">523</span>  <span class="m">261</span>  <span class="m">555</span>  <span class="m">976</span>  <span class="m">500</span>  <span class="m">787</span>  <span class="m">507</span>  <span class="m">440</span>  <span class="m">497</span>  <span class="m">832</span>  <span class="m">529</span>
</pre></div>
</div>
<p>Note that by default the count() function includes all overlapping DNA words in
a sequence. Therefore, for example, the sequence &#8220;ATG&#8221; is considered to contain
two words that are two nucleotides long: &#8220;AT&#8221; and &#8220;TG&#8221;.</p>
<p>If you type help(&#8216;count&#8217;), you will see that the result (output) of
the function count() is a <em>table</em> object. This means that you can
use double square brackets to extract the values of elements from
the table. For example, to extract the value of the third element
(the number of Gs in the DEN-1 Dengue virus sequence), you can type:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> denguetable <span class="o">&lt;-</span> count<span class="p">(</span>dengueseq<span class="p">,</span><span class="m">1</span><span class="p">)</span>
<span class="o">&gt;</span> denguetable<span class="p">[[</span><span class="m">3</span><span class="p">]]</span>
 <span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">2770</span>
</pre></div>
</div>
<p>The command above extracts the third element of the table produced
by count(dengueseq,1), which we have stored in the table variable
<em>denguetable</em>.</p>
<p>Alternatively, you can find the value of the element of the table
in column &#8220;g&#8221; by typing:</p>
<div class="highlight-r"><div class="highlight"><pre><span class="o">&gt;</span> denguetable<span class="p">[[</span><span class="s">&quot;g&quot;</span><span class="p">]]</span>
 <span class="p">[</span><span class="m">1</span><span class="p">]</span> <span class="m">2770</span>
</pre></div>
</div>
</div>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="chapter1.html#summary" title="Permalink to this headline">¶</a></h2>
<p>In this practical, you will have learnt to use the following R
functions:</p>
<ol class="arabic simple">
<li>length() for finding the length of a vector or list</li>
<li>table() for printing out a table of the number of occurrences of
each type of item in a vector or list.</li>
</ol>
<p>These functions belong to the standard installation of R.</p>
<p>You have also learnt the following R functions that belong to the
SeqinR package:</p>
<ol class="arabic simple">
<li>GC() for calculating the GC content for a DNA sequence</li>
<li>count() for calculating the number of occurrences of DNA words
of a particular length in a DNA sequence</li>
</ol>
</div>
<div class="section" id="links-and-further-reading">
<h2>Links and Further Reading<a class="headerlink" href="chapter1.html#links-and-further-reading" title="Permalink to this headline">¶</a></h2>
<p>Some links are included here for further reading.</p>
<p>For background reading on DNA sequence statistics, it is
recommended to read Chapter 1 of
<em>Introduction to Computational Genomics: a case studies approach</em>
by Cristianini and Hahn (Cambridge University Press;
<a class="reference external" href="http://www.computational-genomics.net/book/">www.computational-genomics.net/book/</a>).</p>
<p>For more in-depth information and more examples on using the SeqinR
package for sequence analysis, look at the SeqinR documentation,
<a class="reference external" href="http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng">http://pbil.univ-lyon1.fr/software/seqinr/doc.php?lang=eng</a>.</p>
<p>There is also a very nice chapter on &#8220;Analyzing Sequences&#8221;, which
includes examples of using SeqinR for sequence analysis, in the
book <em>Applied statistics for bioinformatics using R</em> by Krijnen
(available online at
<a class="reference external" href="http://cran.r-project.org/doc/contrib/Krijnen-IntroBioInfStatistics.pdf">cran.r-project.org/doc/contrib/Krijnen-IntroBioInfStatistics.pdf</a>).</p>
<p>For a more in-depth introduction to R, a good online tutorial is
available on the &#8220;Kickstarting R&#8221; website,
<a class="reference external" href="http://cran.r-project.org/doc/contrib/Lemon-kickstart/">cran.r-project.org/doc/contrib/Lemon-kickstart</a>.</p>
<p>There is another nice (slightly more in-depth) tutorial to R
available on the &#8220;Introduction to R&#8221; website,
<a class="reference external" href="http://cran.r-project.org/doc/manuals/R-intro.html">cran.r-project.org/doc/manuals/R-intro.html</a>.</p>
</div>
<div class="section" id="acknowledgements">
<h2>Acknowledgements<a class="headerlink" href="chapter1.html#acknowledgements" title="Permalink to this headline">¶</a></h2>
<p>Thank you to Noel O&#8217;Boyle for helping in using Sphinx, <a class="reference external" href="http://sphinx.pocoo.org">http://sphinx.pocoo.org</a>, to create
this document, and github, <a class="reference external" href="https://github.com/">https://github.com/</a>, to store different versions of the document
as I was writing it, and readthedocs, <a class="reference external" href="http://readthedocs.org/">http://readthedocs.org/</a>, to build and distribute
this document.</p>
<p>Many of the ideas for the examples and exercises for this chapter
were inspired by the Matlab case studies on
<em>Haemophilus influenzae</em>
(<a class="reference external" href="http://www.computational-genomics.net/case_studies/haemophilus_demo.html">www.computational-genomics.net/case_studies/haemophilus_demo.html</a>)
and Bacteriophage lambda
(<a class="reference external" href="http://www.computational-genomics.net/case_studies/lambdaphage_demo.html">http://www.computational-genomics.net/case_studies/lambdaphage_demo.html</a>)
from the website that accompanies the book
<em>Introduction to Computational Genomics: a case studies approach</em>
by Cristianini and Hahn (Cambridge University Press;
<a class="reference external" href="http://www.computational-genomics.net/book/">www.computational-genomics.net/book/</a>).</p>
<p>Thank you to Jean Lobry and Simon Penel for helpful advice on using
the SeqinR package.</p>
</div>
<div class="section" id="contact">
<h2>Contact<a class="headerlink" href="chapter1.html#contact" title="Permalink to this headline">¶</a></h2>
<p>I will be grateful if you will send me (<a class="reference external" href="http://www.sanger.ac.uk/research/projects/parasitegenomics/">Avril Coghlan</a>) corrections or suggestions for improvements to
my email address <a class="reference external" href="mailto:alc&#37;&#52;&#48;sanger&#46;ac&#46;uk">alc<span>&#64;</span>sanger<span>&#46;</span>ac<span>&#46;</span>uk</a></p>
</div>
<div class="section" id="license">
<h2>License<a class="headerlink" href="chapter1.html#license" title="Permalink to this headline">¶</a></h2>
<p>The content in this book is licensed under a <a class="reference external" href="http://creativecommons.org/licenses/by/3.0/">Creative Commons Attribution 3.0 License</a>.</p>
</div>
<div class="section" id="exercises">
<h2>Exercises<a class="headerlink" href="chapter1.html#exercises" title="Permalink to this headline">¶</a></h2>
<p>Answer the following questions, using the R package. For each
question, please record your answer, and what you typed into R to
get this answer.</p>
<p>Model answers to the exercises are given in
<a class="reference external" href="chapter_answers.html#dna-sequence-statistics-1">Answers to the exercises on DNA Sequence Statistics (1)</a>.</p>
<p>Q1. What are the last twenty nucleotides of the Dengue virus genome sequence?</p>
<dl class="docutils">
<dt>Q2. What is the length in nucleotides of the genome sequence for the bacterium <em>Mycobacterium leprae</em> strain TN (accession NC_002677)?</dt>
<dd>Note: <em>Mycobacterium leprae</em> is a bacterium that is responsible for causing
<a class="reference external" href="http://apps.who.int/tdr/svc/diseases/leprosy">leprosy</a>,
which is classified by the WHO as a neglected tropical disease.
As the genome sequence is a DNA sequence, if you are retrieving its sequence via the NCBI website,
you will need to look for it in the NCBI Nucleotide database.</dd>
<dt>Q3. How many of each of the four nucleotides A, C, T and G, and any other symbols, are there in the <em>Mycobacterium leprae</em> TN genome sequence?</dt>
<dd>Note: other symbols apart from the four nucleotides A/C/T/G may
appear in a sequence. They correspond to positions in the sequence
that are are not clearly one base or another and they are due, for
example, to sequencing uncertainties. or example, the symbol &#8216;N&#8217;
means &#8216;aNy base&#8217;, while &#8216;R&#8217; means &#8216;A or G&#8217; (puRine). There is a
table of symbols at
<a class="reference external" href="http://www.bioinformatics.org/sms/iupac.html">www.bioinformatics.org/sms/iupac.html</a>.</dd>
<dt>Q4. What is the GC content of the <em>Mycobacterium leprae</em> TN genome sequence, when (i) all non-A/C/T/G nucleotides are included, (ii) non-A/C/T/G nucleotides are discarded?</dt>
<dd>Hint: look at the help page for the GC() function to find out how
it deals with non-A/C/T/G nucleotides.</dd>
<dt>Q5. How many of each of the four nucleotides A, C, T and G are there in the complement of the <em>Mycobacterium leprae</em> TN genome sequence?</dt>
<dd>Hint: you will first need to search for a function to calculate the
complement of a sequence. Once you have found out what function to
use, remember to use the help() function to find out what are the
arguments (inputs) and results (outputs) of that function. How does
the function deal with symbols other than the four nucleotides A,
C, T and G?
Are the numbers of As, Cs, Ts, and Gs in the complementary sequence
what you would expect?</dd>
</dl>
<p>Q6. How many occurrences of the DNA words CC, CG and GC occur in the <em>Mycobacterium leprae</em> TN genome sequence?</p>
<dl class="docutils">
<dt>Q7. How many occurrences of the DNA words CC, CG and GC occur in the (i) first 1000 and (ii) last 1000 nucleotides of the <em>Mycobacterium leprae</em> TN genome sequence?</dt>
<dd>How can you check that the subsequence that you have looked at is
1000 nucleotides long?</dd>
</dl>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="chapter2.html" title="DNA Sequence Statistics (2)"
             >next</a> |</li>
        <li class="right" >
          <a href="chapter0.html" title="Neglected Tropical diseases"
             >previous</a> |</li>
        <li><a href="https://a-little-book-of-r-for-bioinformatics.readthedocs.org/en/latest/index.html">Bioinformatics 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright Avril Coghlan, Licence: Creative Commons Attribution 3.0 License.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>